go/vt/sidecardb/sidecardb.go

```diff
@@ -202,6 +208,31 @@ func Init(ctx context.Context, exec Exec) error {
 	return nil
 }
 
+// We need to allow zero dates for existing sidecar database tables which may happen to
+// actually have zero (in) date values. setSQLMode gets the current sql_mode, change it to a more relaxed value,
+// defer restoring it to the original value.
+func (si *schemaInit) setSQLMode() (func(), error) {
+	rs, err := si.exec(si.ctx, `select @@session.sql_mode as sql_mode`, 1, false)
+	if err != nil {
+		return nil, vterrors.Errorf(vtrpcpb.Code_UNKNOWN, "could not read sql_mode: %v", err)
+	}
+	sqlMode, err := rs.Named().Row().ToString("sql_mode")
+	if err != nil {
+		return nil, vterrors.Errorf(vtrpcpb.Code_UNKNOWN, "could not read sql_mode: %v", err)
+	}
+
+	resetSQLModeFunc := func() {
+		restoreSQLModeQuery := fmt.Sprintf("set @@session.sql_mode='%s'", sqlMode)
+		_, _ = si.exec(si.ctx, restoreSQLModeQuery, 0, false)
+	}
+
+	changeSQLModeQuery := fmt.Sprintf("set @@session.sql_mode=REPLACE(REPLACE('%s', 'NO_ZERO_DATE', ''), 'NO_ZERO_IN_DATE', '')", sqlMode)
```

Comment:
I think this should be more general. IMO, the point here is that any data that may currently exist in the table — however it got there (keep in mind that `SQL_MODE` is a session level setting) — should not prevent us from reaching the desired schema state. 

I think that we should become fully permissive here (`set sql_mode=''`) as we are not concerned with the data in the table at this point but rather the schema (and things like this do not violate the schema).

Reply:
I wasn't sure if doing that would suppress schema-related errors. So it is guaranteed that `sql_mode` only applies to data validations?

Reply:
Trying to think of other possible scenarios -- no zero for auto-increment? What else could fail the migration?
I'm happy to set an empty `sql_mode`.

> as we are not concerned with the data in the table at this point but rather the schema 

Just pointing out the the bug this PR solves has to do with the data, not with the schema. We';re solving an issue where a row has a timestamp value of `0000-00-00 00:00:00`

Reply:
The DDL defines the table's schema. The `sidecardb` packages purpose is to ensure that the database is using that schema. `SQL_MODE` impacts the MySQL server's behavior beyond the defined schema — the SQL syntax MySQL supports and the data validation checks it performs beyond what the table defines. In our case, disabling the `SQL_MODE` settings ensures that we can apply the correct schema regardless of what data is *already* in the table. 0 dates are merely one example today.

Reply:
To be clear, I can't think of any specific other failure that we'd encounter today. My point is more general and forward looking (we should address the fundamental issue rather than the specific case encountered today).

Reply:
fixed